### PR TITLE
Add materil xml file to UFP file

### PR DESCRIPTION
**Pull request by alekseisasin**
CURA-5555

### CODE REVIEW

File: plugins/UFPWriter/UFPWriter.py [hunk 1]
```diff
@@ -60,5 +60,43 @@ def write(self, stream, nodes, mode = MeshWriter.OutputMode.BinaryMode):
         else:
             Logger.log("d", "Thumbnail not created, cannot save it")
 
+        #Store the material.
+        application = Application.getInstance()
+        machine_manager = application.getMachineManager()
+        global_stack = machine_manager.activeMachine
+        container_registry = application.getContainerRegistry()
+
+        archive.addContentType(extension="xml.fdm_material", mime_type="application/x-ultimaker-material-profile")
```
Comment (ChrisTerBeke) [inline plugins/UFPWriter/UFPWriter.py]: If the content type was already added to the archive this call will throw an exception. Even though it shouldn't already have been added, you never know (defensive programming).

File: plugins/UFPWriter/UFPWriter.py [hunk 2]
```diff
@@ -60,5 +60,43 @@ def write(self, stream, nodes, mode = MeshWriter.OutputMode.BinaryMode):
...
+
+        added_materials = []
+        for extruder_stack in global_stack.extruders.values():
+            material = extruder_stack.material
+            material_file_name = material.getMetaData()["base_file"] + ".xml.fdm_material"
+            material_file_name = "/Cura/" + material_file_name
```
Comment (ChrisTerBeke) [inline plugins/UFPWriter/UFPWriter.py]: What does /Cura do here? (I know it's the (not final?) file path, but it still can use some inline documentation).

File: plugins/UFPWriter/UFPWriter.py [hunk 3]
```diff
@@ -60,5 +60,43 @@ def write(self, stream, nodes, mode = MeshWriter.OutputMode.BinaryMode):
...
+
+            #Same material cannot be added
+            if material_file_name in added_materials:
+                continue
+
+            material_file = archive.getStream(material_file_name)
```
Comment (ChrisTerBeke) [inline plugins/UFPWriter/UFPWriter.py]: There's not point in already opening the stream if in the next step the container cannot be found.

File: plugins/UFPWriter/UFPWriter.py [hunk 4]
```diff
@@ -60,5 +60,43 @@ def write(self, stream, nodes, mode = MeshWriter.OutputMode.BinaryMode):
...
+
+            material_container = material_containers[0]
+            try:
+                serialized_material = material_container.serialize()
+            except NotImplementedError:
+                Logger.log("e", "Unable serialize material container with id: %s", material.id)
+                return False
+
+            material_file.write(serialized_material.encode("UTF-8"))
+            archive.addRelation(virtual_path=material_file_name,relation_type="http://schemas.ultimaker.org/package/2018/relationships/xml.fdm_material")
```
Comment (ChrisTerBeke) [inline plugins/UFPWriter/UFPWriter.py]: Missing the root of the relation here? (e.g. what is this new file related to)

Comment (ChrisTerBeke) [reply plugins/UFPWriter/UFPWriter.py]: This comment is not addressed yet.

**Comment by ChrisTerBeke** [comment]
Still missing (unit) tests.

File: plugins/UFPWriter/UFPWriter.py [hunk 5]
```diff
@@ -60,5 +62,48 @@ def write(self, stream, nodes, mode = MeshWriter.OutputMode.BinaryMode):
...
+            material_container = material_containers[0]
+            try:
+                serialized_material = material_container.serialize()
+            except NotImplementedError:
+                Logger.log("e", "Unable serialize material container with id: %s", material.id)
+                return False
+
+            material_file = archive.getStream(material_file_name)
+            material_file.write(serialized_material.encode("UTF-8"))
+            archive.addRelation(virtual_path=material_file_name,relation_type="http://schemas.ultimaker.org/package/2018/relationships/xml.fdm_material")
```
Comment (LipuFei) [inline plugins/UFPWriter/UFPWriter.py]: - Use `"http://schemas.ultimaker.org/package/2018/relationships/material"` for `relation_type` 
- Add `origin` to use the gcode file.

File: plugins/UFPWriter/UFPWriter.py [hunk 6]
```diff
@@ -24,6 +24,8 @@ def __init__(self):
         self._snapshot = None
         Application.getInstance().getOutputDeviceManager().writeStarted.connect(self._createSnapshot)
 
+        self._default_data_path ="/Cura/"
```
Comment (LipuFei) [inline plugins/UFPWriter/UFPWriter.py]: Put the materials in `/Materials`

**Comment by LipuFei** [comment]
@ChrisTerBeke Discussed with @nallath and updated this PR. For unit testing, I don't think it is possible right now because of all the dependencies.

**Comment by ChrisTerBeke** [comment]
Sure, your call. Be sure to merge master into here first though so we can see if the CI succeeds.

**Pull request merged**